pkg/filters/redirector/redirector.go

```diff
@@ -0,0 +1,174 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package redirector
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/filters"
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/protocols/httpprot"
+	"github.com/megaease/easegress/pkg/util/stringtool"
+)
+
+const (
+	// Kind is the kind of Redirector.
+	Kind = "Redirector"
+
+	resultMismatch = "mismatch"
+)
+
+const (
+	matchPartFull = "full"
+	matchPartURI  = "uri"
+	matchPartPath = "path"
+)
+
+var statusCodeMap = map[int]string{
+	301: "Moved Permanently",
+	302: "Found",
+	303: "See Other",
+	304: "Not Modified",
+	307: "Temporary Redirect",
+	308: "Permanent Redirect",
+}
+
+var kind = &filters.Kind{
+	Name:        Kind,
+	Description: "Redirector redirect HTTP requests.",
+	Results:     []string{resultMismatch},
+	DefaultSpec: func() filters.Spec {
+		return &Spec{}
+	},
+	CreateInstance: func(spec filters.Spec) filters.Filter {
+		return &Redirector{spec: spec.(*Spec)}
+	},
+}
+
+func init() {
+	filters.Register(kind)
+}
+
+type (
+	// Redirector is filter to redirect HTTP requests.
+	Redirector struct {
+		spec *Spec
+		re   *regexp.Regexp
+	}
+
+	// Spec describes the Redirector.
+	Spec struct {
+		filters.BaseSpec `json:",inline"`
+
+		Match       string `json:"match" jsonschema:"required"`
+		MatchPart   string `json:"matchPart" jsonschema:"required,enum=uri|full|path"` // default uri
+		Replacement string `json:"replacement" jsonschema:"required"`
+		StatusCode  int    `json:"statusCode" jsonschema:"required,enum=300|301|302|303|304|307|308"` // default 301
+	}
+)
+
+// Name returns the name of the Redirector filter instance.
+func (r *Redirector) Name() string {
+	return r.spec.Name()
+}
+
+// Kind returns the kind of Redirector.
+func (r *Redirector) Kind() *filters.Kind {
+	return kind
+}
+
+// Spec returns the spec used by the Redirector
+func (r *Redirector) Spec() filters.Spec {
+	return r.spec
+}
+
+// Init initializes Redirector.
+func (r *Redirector) Init() {
+	r.reload()
+}
+
+// Inherit inherits previous generation of Redirector.
+func (r *Redirector) Inherit(previousGeneration filters.Filter) {
+	r.Init()
+}
+
+func (r *Redirector) reload() {
+	if r.spec.StatusCode == 0 {
+		r.spec.StatusCode = 301
+	}
+	if _, ok := statusCodeMap[r.spec.StatusCode]; !ok {
+		logger.Warnf("invalid status code of Redirector, support 300, 301, 302, 303, 304, 307, 308, use 301 instead")
```

Comment:
Actually, you can log `logger.Errorf("BUG: ...")`, because it can't get the wrong code here after validating spec.